src/coordinator/tsdb/remote/compressed_codecs.go

```diff
@@ -142,25 +158,52 @@ func CompressedSeriesFromSeriesIterator(it encoding.SeriesIterator) (*rpc.Series
 		})
 	}
 
-	tags, err := compressedTagsFromTagIterator(it.Tags())
-	if err != nil {
-		return nil, err
-	}
-
 	start := xtime.ToNanoseconds(it.Start())
 	end := xtime.ToNanoseconds(it.End())
+	tagIter := it.Tags()
+	compressedTags, canCompress := compressedTagsFromTagIteratorWithEncoder(tagIter, iterPools)
+
+	var tags []*rpc.Tag
+	var err error
+	if !canCompress {
+		tags, err = tagsFromTagIterator(tagIter)
+		if err != nil {
+			return nil, err
+		}
+	}
 
 	compressedDatapoints := &rpc.CompressedDatapoints{
-		Namespace: it.Namespace().Bytes(),
-		StartTime: start,
-		EndTime:   end,
-		Replicas:  compressedReplicas,
-		Tags:      tags,
+		Namespace:      it.Namespace().Bytes(),
+		StartTime:      start,
+		EndTime:        end,
+		Replicas:       compressedReplicas,
+		CompressedTags: compressedTags,
 	}
 
 	return &rpc.Series{
 		Id:         it.ID().Bytes(),
 		Compressed: compressedDatapoints,
+		Tags:       tags,
+	}, nil
+}
+
+// EncodeToCompressedFetchResult encodes SeriesIterators to compressed fetch results
+func EncodeToCompressedFetchResult(
+	iterators encoding.SeriesIterators,
+	iterPools encoding.IteratorPools,
+) (*rpc.FetchResult, error) {
+	iters := iterators.Iters()
```

Comment:
`defer iterators.Close()`?

Reply:
That's going to be the responsibility of the caller, since tags id and namespace bytes are shared between the iterators and *rpc.FetchResult.